src/sentry/integrations/github_enterprise/integration.py

```diff
@@ -0,0 +1,214 @@
+            attrs={
+                'label': "Github App Name",
+                'placeholder': _('sentry-app'),
+            }
+        ))
+        self.fields['client-id'] = forms.CharField(widget=forms.TextInput(
+            attrs={
+                'label': "Github App Client ID",
+                'placeholder': _('1'),
+            }
+        ))
+        self.fields['client-secret'] = forms.CharField(widget=forms.TextInput(
+            attrs={
+                'label': "Github App Client Secret",
+                'placeholder': _('XXXXXXXXXXXXXXXXXXXXXXXXXXX'),
+            }
+        ))
+        self.fields['webhook-secret'] = forms.CharField(required=False, widget=forms.TextInput(
+            attrs={
+                'label': "Github App Webhook Secret",
+                'placeholder': _('XXXXXXXXXXXXXXXXXXXXXXXXXXX'),
+            }
+        ))
+        self.fields['private-key'] = forms.CharField(
+            widget=forms.Textarea(attrs={'rows': '60',
+                                         'label': "Github App Private Key",
+                                         'placeholder': _("""-----BEGIN RSA PRIVATE KEY-----
+XXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXX
+-----END RSA PRIVATE KEY-----
+"""), })
+        )
+
+
+class InstallationConfigView(PipelineView):
+    def dispatch(self, request, pipeline):
+        form = InstallationForm(request.POST)
+        # TODO(maxbittker) handle errors
+        if form.is_valid():
+
+            form_data = form.cleaned_data
+            form_data['url'] = urlparse(form_data['url']).netloc
+
+            pipeline.bind_state('installation_data', form_data)
+
+            pipeline.bind_state('oauth_config_information', {
+                "access_token_url": "https://{}/login/oauth/access_token".format(form_data.get('url')),
+                "authorize_url": "https://{}/login/oauth/authorize".format(form_data.get('url')),
+                "client_id": form_data.get('client-id'),
+                "private_key": form_data.get('private-key'),
+                "client_secret": form_data.get('client-secret'),
+            })
+
+            return pipeline.next_step()
+
+        project_form = InstallationForm()
+
+        return render_to_response(
+            template='sentry/integrations/github-enterprise-config.html',
+            context={
+                'form': project_form,
+            },
+            request=request,
+        )
+
+
+class GitHubEnterpriseIntegrationProvider(GitHubIntegrationProvider):
+    key = 'github-enterprise'
+    name = 'GitHub Enterprise'
+    metadata = metadata
+
+    def get_pipeline_views(self):
+        identity_pipeline_config = {
+            'oauth_scopes': (),
+            'redirect_url': absolute_uri('/extensions/github-enterprise/setup/'),
+        }
+
+        identity_pipeline_view = NestedPipelineView(
```

Comment:
Okay so I realized, because we now allow for _late binding_ of pipeline steps, we can actually do a better job providing the oauth parameters to the identity pipeline through it's *config* object, versus it expecting to find it in a parent pipeline. How about this

If we have a method helper method to construct the identity pipeline view

```py
def _make_identity_pipeline_view(self):
    """
    Make the nested identity provider view. It is important that this view is
    not constructed until we reach this step and the
    ``oauth_config_information`` is available in the pipeline state. This
    method should be late bound into the pipeline vies.
    """

    identity_pipeline_config = dict(
        oauth_scopes=(),
        redirect_url=absolute_uri('/extensions/github-enterprise/setup/'),
        *self.pipeline.fetch_state('oauth_config_information'),
    )

    return NestedPipelineView(
        bind_key='identity',
        provider_key='github-enterprise',
        pipeline_cls=IdentityProviderPipeline,
        config=identity_pipeline_config,
    )
```

Then in the `get_pipeline_views`

```py
return [
    InstallationConfigView(),
    GitHubEnterpriseInstallationRedirect(),

    # The identity provider pipeline should be constructed at execution
    # time, this allows for the oauth configuration parameters to be made
    # available from the installation config view.
    lambda: self._make_identity_pipeline_view(),
]
```